### ISSUE TITLE
Processing jobs queue

### ISSUE THREAD

**Issue by kdarras** [ecomontec/ecoSound-web#125]
We need to have a transparent system to launch, check, and manage processing jobs running on batches of recordings.

The batch execution of resource-hungry tasks should only be accessible to project managers. We can currently execute simple actions on multiple recordings thanks to #342 but demanding ones should also be executable from the admin table and manageable with RabbitMQ.

In the future, actions may be possible at other levels than the recording (tags, collections), so this needs to be implemented sustainably.

Examples of job types include:
- recording upload (#412 - done)
- execution of AI models (#413 - done)
- execution of acoustic index analysis (#367 - done)
- downloading original recordings (#411 - implemented separately)
- running queries (not existing yet. See "Queries" project - to implement separately)

Clicking the button to start a job should display the same window that appears for the same function (if it exists) from the player, prompting for parameters.

However, a message should appear next to the button that confirms launching the job when multiple items are selected: "Launch job for X (number of selected recordings) recordings?".
If the user proceeds, a message should say "Job added to your queue - check status under corresponding tab" (similar to pop-up that users get after saving uploaded recordings).
The progress of the jobs should be visible under a new "Queue" tab in the admin interface.

The queue should display:
- the action type (acoustic index analysis/recording upload/AI model, etc.)
- the progress (number of finished items/total items)
- the status (ongoing (grey)/ finished (green) / failed (red))
- the start and end date-time (+ elapsed time?)
- buttons to abort ongoing jobs (ask for confirmation) and to clear (successfully) finished jobs.

Correspondingly, jobs should be logged (user, job type, start and end time, status, error message) in the database.

The user can then continue browsing normally. Ongoing jobs should run independently of whether users are logged in.

The batch analysis of recordings should not negatively the higher-priority browsing operations from the other users.

**Comment by LiuDilongNJ** [comment]
Still use RabbitMQ queue, will be for each user to perform the operation to create a dynamic queue, at the same time more than one user execution may cause other users can not access.

**Comment by kdarras** [comment]
If:
1. you recommend to continue using RabbitMQ
2. it is still possible for multiple users to add jobs to the queue simultaneously
3. RabbitMQ can indeed be adapted to the job types listed above
Then let's continue using it.
It is OK and understandable that jobs are processed sequentially. As long as the users are not logged out and don't lose their jobs, they can wait for the result.

**Comment by LiuDilongNJ** [comment]
1.Yes.
2.Yes.
3.Yes.
I think 1 is better.

**Comment by kdarras** [comment]
Alright. Solution 1 is actually similar to what was implemented on early versions of the website. You might find inspiration in some old BioSounds or Pumilio code.

**Comment by kdarras** [comment]
Question: would the requirement "Ongoing jobs should run independently of whether users are logged in." still work for the upload function? I imagine that if users are logged out, local files cannot upload, right? Would it be better to then add a warning that upload will be interrupted if the user logs out while the job is running, and to extend the login time until the job is completed?

**Comment by kdarras** [comment]
When failures happen, should we report which items failed and save the ouput of the successful ones (preferred solution) or roll-back if any failure happens?

**Comment by LiuDilongNJ** [comment]
The uploading of recordings is fast, the formatting and generation of spectrograms is slow, and this is what the background queue has to deal with, even if the user is logged off, it can still accomplish its task.

When the queue fails, the successes and failures are labeled.

**Comment by kdarras** [comment]
Let's use the singular "Queue" (we have only one queue and it can contain multiple items) for naming the tab in the admin interface
couldn't test yet as I could not upload recordings yet

**Comment by LiuDilongNJ** [comment]
Later I'll set it to refresh every 5 seconds if there are tasks in progress. If a user submits a new task to the queue, we'll need to manually refresh the page or set a refresh button or change it to automatically refresh the page every 15 seconds. Also, regarding the colour of the states, currently I have set 5 states, because when terminating a queue, it will go into a cancelled state and wait for the tasks in the queue to finish before turning into a cancelled state.
![image](https://github.com/ecomontec/ecoSound-web/assets/109188679/844b6e6c-b65a-4162-b302-a6a93882e07f)

**Comment by kdarras** [comment]
Thanks. Not sure I understood the two "cancelled" statuses and 5/7 states, but the periodic automatic refreshes of the page sound like a good solution to me.

**Comment by kdarras** [comment]
The job queue works beautifully.

Some issues remain:
- checked boxes are unchecked with every automatic refresh, resulting in an error if one continues:
![image](https://github.com/ecomontec/ecoSound-web/assets/7696198/a037bc4b-c500-4a58-bb76-8c7f4d57dd46)
- the message "Select at least one audio data." Should be "Select at least one audio recording."

I will stress-test the feature on a new collection of 3000 recordings

**Comment by kdarras** [comment]
Works in Docker

**Comment by kdarras** [comment]
I noticed that when two jobs are running, they are both shown as "ongoing" even though one should be "pending":
![image](https://github.com/ecomontec/ecoSound-web/assets/7696198/ec0073c5-4baf-4a19-b99f-2271efc25e77)
Could this be adjusted?

**Comment by kdarras** [comment]
Note that in the same session, this warning also appeared:
![image](https://github.com/ecomontec/ecoSound-web/assets/7696198/aeb4bf0b-e3f8-4a4a-927e-8f64fdb2498e)

But the job in the queue proceeded.
It looks similar to what I had when using "show all entries" for tables with many items, and I had indeed just used it on the recordings table.

So I returned to the recordings table to set it back to "show 10 entries" and then I returned to the queue tab. I was surprised to see this error:
![image](https://github.com/ecomontec/ecoSound-web/assets/7696198/260878da-37dc-4f61-8285-aeb855d7c8db)
Still, the queue job proceeded and the warning vanished after a page refresh.
